thewalrus/_hermite_multidimensional.py

```diff
@@ -143,4 +148,160 @@ def hafnian_batched(A, cutoff, mu=None, rtol=1e-05, atol=1e-08, renorm=False, ma
     return hermite_multidimensional(
         -A, cutoff, y=mu, renorm=renorm, make_tensor=make_tensor, modified=True
     )
+
+
 # Note the minus signs in the arguments. Those are intentional and are due to the fact that Dodonov et al. in PRA 50, 813 (1994) use (p,q) ordering instead of (q,p) ordering
+
+
+@jit(nopython=True)
+def dec(tup: Tuple[int], i: int) -> Tuple[int, ...]:  # pragma: no cover
+    r"""returns a copy of the given tuple of integers where the ith element has been decreased by 1
+
+    Args:
+        tup (Tuple[int]): the given tuple
+        i (int): the position of the element to be decreased
+
+    Returns:
+        Tuple[int,...]: the new tuple with the decrease on i-th element by 1
+    """
+    copy = tup[:]
+    return tuple_setitem(copy, i, tup[i] - 1)
+
+
+@jit(nopython=True)
+def remove(
+    pattern: Tuple[int, ...]
+) -> Generator[Tuple[int, Tuple[int, ...]], None, None]:  # pragma: no cover
+    r"""returns a generator for all the possible ways to decrease elements of the given tuple by 1
+
+    Args:
+        pattern (Tuple[int, ...]): the pattern given to be decreased
+
+    Returns:
+        Generator[Tuple[int, Tuple[int, ...]], None, None]: the generator
+    """
+    for p, n in enumerate(pattern):
+        if n > 0:
+            yield p, dec(pattern, p)
+
+
+SQRT = np.sqrt(np.arange(1000))  # saving the time to recompute square roots
+
+
+def hermite_multidimensional_numba(C, R, cutoff, y, dtype=np.complex128):
```

Comment:
my final suggestion is to make the new functions follow as close as possible the signature of the other functions in the module, which could be done by moving `C` from being the first (and mandatory) argument of the function to be an optional argument at the end of the list of arguments.